### ISSUE TITLE
SimCLR distributed training

### ISSUE THREAD

**Issue by jlindsey15** [open-mmlab/mmselfsup#3]
Hi!  Thanks for this awesome repo.  The official SimCLR implementation (https://github.com/google-research/simclr) does not provide a multi-gpu implementation "for reasons such as global BatchNorm and contrastive loss across cores."  Does this implementation solve that issue, and if so, does it attain the same performance as the official code/model?

**Comment by XiaohangZhan** [comment]
This implementation supports multi-gpus distributed training (DistributedDataParallel). I used the PyTorch official SyncBatchNorm as the global batchnorm. The contrastive loss across cores is enabled by a GatherLayer (https://github.com/open-mmlab/OpenSelfSup/blob/696d04950e55d504cf33bc83cfadbb4ece10fbae/openselfsup/models/simclr.py#L55) that gathers tensors across GPUs and supports backward propagation. The official SimCLR implementation in TensorFlow should support distributed training, I'm not sure the differences between "distributed training" and "multi-GPUs training" in TensorFlow.

Regarding the performance, due to the computation resources, I can merely run a SimCLR with batchsize=256, epoch=200. The linear ImageNet classification evaluation is ongoing.

A table comparing the performance of different methods in all benchmarks will appear in recent days.

**Comment by jlindsey15** [comment]
Thank you!

**Comment by Quentinzzz** [comment]
> This implementation supports multi-gpus distributed training (DistributedDataParallel). I used the PyTorch official SyncBatchNorm as the global batchnorm. The contrastive loss across cores is enabled by a GatherLayer (
> 
> https://github.com/open-mmlab/OpenSelfSup/blob/696d04950e55d504cf33bc83cfadbb4ece10fbae/openselfsup/models/simclr.py#L55
> 
> ) that gathers tensors across GPUs and supports backward propagation. The official SimCLR implementation in TensorFlow should support distributed training, I'm not sure the differences between "distributed training" and "multi-GPUs training" in TensorFlow.
> Regarding the performance, due to the computation resources, I can merely run a SimCLR with batchsize=256, epoch=200. The linear ImageNet classification evaluation is ongoing.
> 
> A table comparing the performance of different methods in all benchmarks will appear in recent days.

Hi, thanks for your great implementation!
I have tried the **GatherLayer** in my own model and task (BERT for NLP), but run into some strange memory issues.

- w/o the GatherLayer operation (in this case, the loss is computed within each GPU device, and there is no cross GPU communication.), the memory cosumption is **6423/6785/7355/7803/8485 MB** for batchsize per GPU of **16/20/24/28/32**, this is as expected.

- But when I import and use the GatherLayer to compute a global contrastive loss, the memory consumption increase very sharply: **8271/12265/23275/31691/OOM** for **16/20/24/28/32**, I think it's very strange that the memory would increase from **12265** to **23275** simply because I add 4 more instances per GPU, It is also strange that memory increase from **7803** to **31691** simply because I aggregated the output feature from all GPUs, the gatherlayer is three times large as the entire encoder model... 
so there might be something wrong with GatherLayer memory consumption or the way I adopted it.

Note that batchsize of 16 means that each GPU produce feature of [16x2, 128] size, and I'm training on 8 32G V100 GPUs.

Really appreciate any reply and help, thx!
